New gadget: `sigsnoop` (#507)
kinvolk/inspektor-gadget: docs/guides/sigsnoop.md

weight: 10
---

The `sigsnoop` gadget is used to trace signal sent system-wide.

## How to use it?

First, we need to create one pod for us to play with:

```bash
$ kubectl run debian --image debian:latest sleep inf
```

You can now use the gadget, but output will be empty:

```bash
$ kubectl gadget sigsnoop
NODE             NAMESPACE        POD              CONTAINER        PID    COMM             SIGNAL    TPID   RET
```

Indeed, it is waiting for signals to be sent.
So, in *an other terminal*, `exec` the container and send one signal:

```bash
$ kubectl exec -ti debian -- sh -c 'sleep 3 & kill -kill $!'
```

Go back to *the first terminal* and see:

```
NODE             NAMESPACE        POD              CONTAINER        PID    COMM             SIGNAL    TPID   RET
minikube         default          debian           debian           129484 sh               SIGKILL   129491 0
minikube         default          debian           debian           129484 sh               SIGHUP    129491 0
minikube         default          debian           debian           129484 sh               SIGHUP    129484 0
```

The first line corresponds to `kill` sending signal `SIGKILL` to `sleep`.

> Note that, you can only use this gadget to trace segmentation fault.

Reviewer: ```suggestion
You can also use this gadget to trace when pods die with segmentation fault.
```

> Note that, you can only use this gadget to trace segmentation fault.

Reviewer 2: Do you mean "also" instead of "only"?

In the *other terminal*, `exec` the container with the following:

```bash
$ kubectl exec -ti debian -- bash
# We first need to install python2.
# apt update && apt install -qy python2
# We can now generate a segfault.
# python2.7 -c "exec'()'*7**6"
```

Now, go back to the first terminal and see that `SIGSEGV` was sent to python:

```
minikube         default          debian           debian           142244 python2.7        SIGSEGV   142244 0
```

> ## Restraining output to certain PID, signal or failed sending
> 
> With the following option, you can restrain the output:

Reviewer: ```suggestion
## Restricting output to certain PIDs, signals or failure to send the signals

With the following option, you can restrict the output:
```

* `--pid` only prints events where a signal is sent by the given PID.
* `--signal` only prints events where the given signal is sent.

> * `-f/--failed-only` only prints events where signal delivery failed.

Reviewer: ```suggestion
* `-f/--failed-only` only prints events where signal failed to be delivered.
```

> So, this command will only prints failed sending of `SIGKILL` sent by PID `42`:

Reviewer: ```suggestion
So, this command will only print failed attempts to send `SIGKILL` by PID `42`:
```

```bash
$ kubectl gadget -f --pid 42 --signal SIGKILL
```

Note that, with `--signal` you can use the name of the signal (e.g. `SIGKILL`) or its integer value (e.g. 9).

## Only print some information

> You can restrain the information printed using `-o custom-columns=column0,...,columnN`.

Reviewer: ```suggestion
You can customize the information printed using `-o custom-columns=column0,...,columnN`.
```

---
title: 'The "sigsnoop" gadget'
weight: 10
---

> The `sigsnoop` gadget is used to traces signal sent system-wide..

Reviewer 3: traces > trace

Reviewer 3: `..` -> `.`?

> In this guide, we will learn how to use it by running a small `kubernetes` cluster inside `minikube`.

Reviewer 3: Why specific for minikube?

NODE             NAMESPACE        POD              CONTAINER        PID    COMM             SIGNAL    TPID   RET
minikube         default          debian           debian           129484 sh               SIGKILL   129491 0
minikube         default          debian           debian           129484 sh               SIGHUP    129491 0
minikube         default          debian           debian           129484 sh               SIGHUP    129484 0
```

## Use JSON output

This gadget supports JSON output, for this simply use `-o json`:

```bash
$ kubectl gadget sigsnoop -o json
{"type":"normal","node":"minikube","namespace":"default","pod":"debian","container":"debian","pid":142872,"tpid":142885,"signal":9,"comm":"sh","mountnsid":4026532588}
# You can use jq to make the output easier to read:
$ kubectl gadget sigsnoop -o json | jq
{
  "type": "normal",
  "node": "minikube",
  "namespace": "default",
  "pod": "debian",
  "container": "debian",
  "pid": 142872,
  "tpid": 142885,
  "signal": 9,
  "comm": "sh",
  "mountnsid": 4026532588
}
```

## Clean everything

Congratulations! You reached the end of this guide!

> You can now delete the two pods we created:

Reviewer 3: two?
